larpandoracontent/LArReclustering/SimplePCAThreeDClusteringTool.cc

```diff
@@ -0,0 +1,120 @@
+/**
+ *  @file   larpandoracontent/LArReclustering/SimplePCAThreeDClusteringTool.cc
+ *
+ *  @brief  Implementation file for the simple PCA-based clustering tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArHelpers/LArPcaHelper.h"
+#include "larpandoracontent/LArReclustering/SimplePCAThreeDClusteringTool.h"
+#include "larpandoracontent/LArHelpers/LArPointingClusterHelper.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+SimplePCAThreeDClusteringTool::SimplePCAThreeDClusteringTool()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+std::vector<std::reference_wrapper<pandora::CaloHitList>> SimplePCAThreeDClusteringTool::Run(const Algorithm *const /*pAlgorithm*/, std::reference_wrapper<pandora::CaloHitList> &inputCaloHitList)
+{
+    std::vector<std::reference_wrapper<pandora::CaloHitList>> newCaloHitListsVector;
+
+    // Begin with a PCA
+    CartesianVector centroid(0.f, 0.f, 0.f);
+    LArPcaHelper::EigenVectors eigenVecs;
+    LArPcaHelper::EigenValues eigenValues(0.f, 0.f, 0.f);
+    LArPcaHelper::RunPca(inputCaloHitList.get(), centroid, eigenValues, eigenVecs);
+
+    // By convention, the primary axis has a positive z-component.
+    const CartesianVector axisDirection(eigenVecs.at(0).GetZ() > 0.f ? eigenVecs.at(0) : eigenVecs.at(0) * -1.f);
+
+    // Place intercept at hit with minimum projection
+    float minProjection(std::numeric_limits<float>::max());
+    for (const CaloHit *const pCaloHit3D : inputCaloHitList.get())
+        minProjection = std::min(minProjection, axisDirection.GetDotProduct(pCaloHit3D->GetPositionVector() - centroid));
+
+    const CartesianVector axisIntercept(centroid + (axisDirection * minProjection));
+
+    // Now define ortho directions
+    const CartesianVector seedDirection((axisDirection.GetX() < std::min(axisDirection.GetY(), axisDirection.GetZ())) ? CartesianVector(1.f, 0.f, 0.f) :
+        (axisDirection.GetY() < std::min(axisDirection.GetX(), axisDirection.GetZ())) ? CartesianVector(0.f, 1.f, 0.f) : CartesianVector(0.f, 0.f, 1.f));
+    const CartesianVector orthoDirection1(seedDirection.GetCrossProduct(axisDirection).GetUnitVector());
+
+    //Lists for hits that have a positive and negative projection on the secondary axis
+    CaloHitList posCaloHitList, negCaloHitList;
+
+    for (const CaloHit *const pCaloHit : inputCaloHitList.get())
+    {
+        const CartesianVector pCaloHitPosition = pCaloHit->GetPositionVector();
+
+        if((pCaloHitPosition-centroid).GetDotProduct(centroid+orthoDirection1)<0)
```

Comment:
maybe i'm missing something.. but why do you need to add the centroid position to the orthodirection? arent you basically saying 'okay how much does the displacement vector of my hit from the centroid lie in the direction of my centroid from the origin?' which i think it not what you want to be asking.. but then of course you also need to make sure that  orthoDirection1 is a unit vector

Reply:
Just a bug, that was not producing visible issues... Thanks for spotting!!